diff --git a/core/index.js b/core/index.js
index fa252be729528754c800641883f0017aab7c50fc..1f27d134b9d9f19d512844bda92568b4cbc58cd6 100644
--- a/core/index.js
+++ b/core/index.js
@@ -55,6 +55,14 @@ async function toInternalRequest(req) {
     };
   }
 
+  if (!req.origin && req.headers) {
+    const {headers} = req
+    req.origin = _detectHost.detectHost(
+      headers["x-forwarded-host"] ?? headers.host,
+      headers["x-forwarded-proto"]
+    )
+  }
+
   return req;
 }
 
